File: python/matrix/matrix_multiplication.py
import random


def generate_matrices(n):
    A = [[random.random() for _ in range(n)] for _ in range(n)]
    B = [[random.random() for _ in range(n)] for _ in range(n)]
    return A, B


def matrix_multiply_standard(A, B):
    n = len(A)
    C = [[0 for _ in range(n)] for _ in range(n)]

    for i in range(n):
        for j in range(n):
            for k in range(n):
                C[i][j] += A[i][k] * B[k][j]
    return C


def matrix_multiply_row_oriented(A, B):
    n = len(A)
    C = [[0 for _ in range(n)] for _ in range(n)]

    for i in range(n):
        for k in range(n):
            for j in range(n):
                C[i][j] += A[i][k] * B[k][j]
    return C


def matrix_multiply_tiled(A, B, block_size=32):
    n = len(A)
    C = [[0 for _ in range(n)] for _ in range(n)]

    for i_block in range(0, n, block_size):
        for j_block in range(0, n, block_size):
            for k_block in range(0, n, block_size):

                i_limit = min(i_block + block_size, n)
                j_limit = min(j_block + block_size, n)
                k_limit = min(k_block + block_size, n)

                for i in range(i_block, i_limit):
                    for k in range(k_block, k_limit):
                        for j in range(j_block, j_limit):
                            C[i][j] += A[i][k] * B[k][j]
    return C

File: code/java/src/main/java/com/pabcablan/matrix/MatrixMultiplication.java
package com.pabcablan.matrix;
import java.util.Random;

public class MatrixMultiplication {
    private final int defaultBlockSize;
    private final Random random;


    public MatrixMultiplication(int defaultBlockSize) {
        this.defaultBlockSize = defaultBlockSize;
        this.random = new Random();
    }


    public double[][][] generateMatrices(int n) {
        double[][] A = new double[n][n];
        double[][] B = new double[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                A[i][j] = random.nextDouble();
                B[i][j] = random.nextDouble();
            }
        }

        return new double[][][] {A, B};
    }


    public double[][] standard(double[][] A, double[][] B) {
        int n = A.length;
        double[][] C = new double[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                for (int k = 0; k < n; k++) {
                    C[i][j] += A[i][k] * B[k][j];
                }
            }
        }

        return C;
    }


    public double[][] rowOriented(double[][] A, double[][] B) {
        int n = A.length;
        double[][] C = new double[n][n];

        for (int i = 0; i < n; i++) {
            for (int k = 0; k < n; k++) {
                for (int j = 0; j < n; j++) {
                    C[i][j] += A[i][k] * B[k][j];
                }
            }
        }

        return C;
    }


    public double[][] tiled(double[][] A, double[][] B, Integer blockSize) {
        if (blockSize == null) {
            blockSize = this.defaultBlockSize;
        }

        int n = A.length;
        double[][] C = new double[n][n];

        for (int iBlock = 0; iBlock < n; iBlock += blockSize) {
            for (int jBlock = 0; jBlock < n; jBlock += blockSize) {
                for (int kBlock = 0; kBlock < n; kBlock += blockSize) {

                    int iLimit = Math.min(iBlock + blockSize, n);
                    int jLimit = Math.min(jBlock + blockSize, n);
                    int kLimit = Math.min(kBlock + blockSize, n);

                    for (int i = iBlock; i < iLimit; i++) {
                        for (int k = kBlock; k < kLimit; k++) {
                            double aik = A[i][k];
                            for (int j = jBlock; j < jLimit; j++) {
                                C[i][j] += aik * B[k][j];
                            }
                        }
                    }
                }
            }
        }

        return C;
    }
}